Split off metadata

jrfnl opened the pull request:
As - unlike with the `Headers` class -, the static methods in the Package class can not easily be overloaded, I'd suggest splitting them off into a separate `Metadata` class which can be overloaded.

This will make it a lot easier to add additional information to the metadata for a package.
All you need to do is extend the Metadata class and overload the $cacheTime $headerMap, $readmeMap and/or constructor to add additional data.
You will of course also need to overload the Package::fromArchive() method to use your extended class rather than the original, but as it's now a small method, it doesn't break the upgrade path as much as before.

Simple example code for how this could now be overloaded:
```php
class My_Metadata extends Wpup_Metadata {

	protected $headerMapExtra = array(
		'Text Domain' => 'textdomain',
	);

	protected $readmeMapExtra = array(
		'short_description',
	);

	public function __construct( $slug, $filename, Wpup_Cache $cache = null ) {
		$this->headerMap = array_merge( $this->headerMap, $this->headerMapExtra );
		$this->readmeMap = array_merge( $this->readmeMap, $this->readmeMapExtra );

		parent::__construct( $slug, $filename, $cache );
	}

	protected function extractMetadata() {
		$metadata = parent::extractMetadata();
		if ( ! is_array( $metadata ) ) {
			return null;
		}

		// Add additional metadata
		$metadata['type'] = $this->packageInfo['type'];

		return $metadata;
	}
}
```

jrfnl commented:
Sorry about not doing anything with this, but I keep struggling with what you propose.

The choice for sticking with static methods and even adding static properties in this case is IMHO just wrong.

The split into two classes is justified as - as documented in the original class - metadata can be retrieved from other sources (db/config file). The implementation I propose(d) in this PR makes it simple to either swap out the `Wpup_Metadata` class for a completely custom implementation for retrieving the metadata or to overload it like shown in my original post - all while maintaining full backward compatibility.

Your suggested implementation on the other hand would make it more complicated rather than simpler.

Sorry, but I stand by my original PR.

YahnisElsts commented:
Nitpick: Still, what's the problem with static methods and static properties? Is it a style issue? Why is it better to use instance methods in this case?

jrfnl commented:
Static methods and properties have their place in programming, just not in this situation.
@mathiasverraes has explained this much better than I could in [this blogpost](http://verraes.net/2014/06/when-to-use-static-methods-in-php/).
What it comes down to is that the methods in the PR could have widely different implementations (taking the data from a database, config files, which metadata will be available) and are therefore not stateless nor entirely predictable.

Here are some more articles which argue in a reasonably balanced manner when to use static and when not (even though the titles are misleading):
http://simpleprogrammer.com/2010/01/29/static-methods-will-shock-you/
http://www.phparch.com/2010/03/static-methods-vs-singletons-choose-neither/
http://kore-nordmann.de/blog/0103_static_considered_harmful.html

jrfnl commented:
@YahnisElsts Thank you for taking the time and your detailed feedback.

> Someone who needs to load metadata from a different source probably wouldn't gain much by resuing it, and would instead either write a completely independent implementation or subclass Wpup_Package.

In which case that custom implementation is now easier to call what with the `$packageFileLoader` property in the UpdateServer class which I implemented based on your suggestion.

> 1. Rename Wpup_Metadata to something like Wpup_ZipMetadataParser

Done.

> 2. Make a general Wpup_Metadata class that only contains stuff that would apply to all packages.

I considered it, but as you say it would be **_extremely_** bare, so I left that out for now until there turns out to be a call for it. Would be easy enough to implement anyway.

> Why is setMetadataFromArchive() public? It's not called from anywhere outside the class.

Artefact from the original static implementation. Fixed.

> This bit looks suspicious to me: ... <snip>... Now it's obvious that $meta is always an array, and that the methods add something to that array.

As it's now a class and the data gets send to the `$metadata` property anyway, I've now refactored the code to use that property everywhere rather than passing the data around all the time.

I've also abstracted out the code a bit more. The original implementation stayed very close to the original methods. I've now refactored it to be more like a proper class ;-)

Hope you approve of the changes.

YahnisElsts commented:
Some thoughts:

Since this is now a class and not a public method, there's no need for extractMetadata() to report an error by setting `$this->metadata` to null. Just let it `throw new Wpup_InvalidPackageException` directly and get rid of the null check in `createMetadataCache`.

On a related note, I disagree with the idea of splitting the cache-related stuff into `setMetadataFromCache` and `createMetadataCache`. Caching is an optional optimization layer. I think it should be as unobtrusive as possible and most code shouldn't know or care about the cache. Calling the main entry point to the parser logic `setMetadataFromCache` gives caching far too much weight. 

Also, in my experience, this is a fairly common pattern when working with cache:

- Try to retrieve data from the cache.
    - On success, return it and exit.
- Generate data the slow way.
- Store it in the cache.

These are all closely related operations. Why split off the last two?

Another thing I noticed is that `setHeaderFields` and `setReadmeFields` are very similar. Both of them copy values from one array to another based on a list of keys. It would be possible write a more general method that maps specific keys from an input array to `$this->metadata`. Something like `setMetadataFields($sourceArray, $keyMap)` (names need work).

`setReadmeSections` is also very similar, except it transforms key names and has no filter, accepting all keys/sections. A more complicated version of `setMetadataFields` could handle that as well.

Finally, a minor style issue: in `extractMetadata`, that `else` should look like this: `} else {`.

jrfnl commented:
Thanks for your feedback again. I agree with most of your points and probably would have done those things already if I wasn't working on ten things at the same time ;-)
I've updated the PR incorporating most of your points.

> setReadmeSections is also very similar, except it transforms key names and has no filter, accepting all keys/sections. A more complicated version of setMetadataFields could handle that as well.

I've looked at combining that into the new method as well, but with there being three differences between how things would need to be handled, it would actually result in a lot more code, rather than less.
The three differences being:
* no map, everything is accepted
* the `str_replace()` for the section name
* the sections going into a 'sections' subarray within `$metadata`

Hope we got everything now.

YahnisElsts commented on includes/Wpup/ZipMetadataParser.php:
```diff
@@ -0,0 +1,239 @@
+	 * @param array $map   The key mapping for that sub-array where the key is the key as used in the
+	 *                     input array and the value is the key to use for the output array
+	 */
+	protected function setMappedFields( $input, $map ){
```
No spaces around method parameters, please.

jrfnl replied:
Fixed, sorry, too used to using the WP coding standards.

YahnisElsts commented:
> I've looked at combining that into the new method as well, but with there being three differences between how things would need to be handled, it would actually result in a lot more code, rather than less.

In theory, you could make the mapping function handle the differences by making it more general and extracting filtering/key mapping logic to a separate method or closure:

```php
protected function mapFields($input, $map, $transform){
	$output = array();
	foreach($input as $fieldKey => $value) {
		$metaKey = call_user_func($transform, $fieldKey, $map);
		if ( $metaKey !== null ) {
			$output[$metaKey] = $value;
		}
	}
	return $output;
}

/* This is the default $transform */
protected function getMappedKey($fieldKey, $map) {
	return isset($map[$fieldKey]) ? $map[$fieldKey] : null;
}

/*...*/

$this->metadata['sections'] = $this->mapFields(
	$this->packageInfo['readme']['sections'],
	array(),
	function($sectionName, $dummy) {
		return str_replace(' ', '_', strtolower($sectionName));
	}
);
```

... but it is indeed more code.

jrfnl commented:
@YahnisElsts So are we good to go now ?

jrfnl commented:
Thank you!
